### PR TITLE
[XPU][Feature] fp8 online quantization support for XPU

### PR DESCRIPTION
**Pull request by yma11**
## Purpose
This PR enable online fp8 quantization on XPU platform with corresponding `dynamic_scaled_fp8_quant` and grouped gemm kernels. Note that need specify `dtype` with `float16` since `bfloat16` has poor perf in current used ipex-2.8 whl used in upstream.

## Test Plan
Verified on several models with `--quantization fp8` added.
```
VLLM_ALLOW_LONG_MAX_MODEL_LEN=1 VLLM_WORKER_MULTIPROC_METHOD=spawn python3 examples/offline_inference/basic/generate.py --model microsoft/Phi-3.5-mini-instruct  --enforce-eager --quantization fp8 --dtype float16
VLLM_ALLOW_LONG_MAX_MODEL_LEN=1 VLLM_WORKER_MULTIPROC_METHOD=spawn python3 examples/offline_inference/basic/generate.py --model meta-llama/Meta-Llama-3-8B-Instruct --enforce-eager --quantization fp8 --dtype float16
VLLM_MLA_DISABLE=1 VLLM_ENABLE_MOE_ALIGN_BLOCK_SIZE_TRITON=1 VLLM_ALLOW_LONG_MAX_MODEL_LEN=1 VLLM_WORKER_MULTIPROC_METHOD=spawn python3 examples/offline_inference/basic/generate.py --model deepseek-ai/DeepSeek-V2-Lite  --enforce-eager --quantization fp8 --dtype float16 -tp=2
```
## Test Result
Benchmark for 1k/512 on intel B60 single card:
```
VLLM_WORKER_MULTIPROC_METHOD=spawn python3 -m vllm.entrypoints.openai.api_server --model meta-llama/Llama-3.1-8B-Instruct --enforce-eager --port 8000 --block-size 64 --gpu-memory-util 0.9 --max-num-batched-tokens=8192 --disable-sliding-window --disable-log-requests --no-enable-prefix-caching --trust-remote-code --max-model-len=4096 --quantization fp8

python3 benchmarks/benchmark_serving.py --model meta-llama/Llama-3.1-8B-Instruct  --dataset-name random --random-input-len=1024 --random-output-len=512 --ignore-eos --num-prompt 130 --request-rate inf --backend vllm --port=8000 --host 0.0.0.0 --trust-remote-code

```
image

acc:
image
## (Optional) Documentation Update

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/_custom_ops.py [hunk 1]
```diff
@@ -1215,8 +1215,12 @@ def scaled_fp8_quant(
             torch.ops._C.dynamic_per_token_scaled_fp8_quant(
                 output, input, scale, scale_ub)
         else:
-            scale = torch.empty(1, device=input.device, dtype=torch.float32)
-            torch.ops._C.dynamic_scaled_fp8_quant(output, input, scale)
+            scale = torch.zeros(1, device=input.device, dtype=torch.float32)
```
Comment (yewentao256) [inline vllm/_custom_ops.py]: I remember one previous PR specifically change to empty, will this cause trouble?

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work!
Could you also post some E2E results? Eg, showing throughput using `vllm bench` and accuracy using `lm-eval`

File: vllm/_custom_ops.py (hunk 1 above)
Comment (yma11) [reply vllm/_custom_ops.py]: Yes. XPU quantization kernel has problem with empty scale. I separate this value to keep empty for non-xpu path, okay for you?

**Comment by yma11** [comment]
> Thanks for the work! Could you also post some E2E results? Eg, showing throughput using `vllm bench` and accuracy using `lm-eval`

benchmark and acc result added, thanks for your review.

File: vllm/_custom_ops.py (hunk 1 above)
Comment (yma11) [reply vllm/_custom_ops.py]: I will do a refactor to put xpu related logic in `_ipex_ops` so there will no change of these code anymore. Will ping you once done.

Comment (yma11) [reply vllm/_custom_ops.py]: Updated.

File: vllm/_ipex_ops.py [hunk 2]
```diff
@@ -349,3 +350,56 @@ def copy_blocks(key_caches: list[torch.Tensor],
...
+                in the dynamic quantization case.
+    
+        Returns:
+            tuple[torch.Tensor, torch.Tensor]: The output tensor in FP8 and
+                scaling factor.
+        """
+        # This code assumes batch_dim and num_tokens are flattened
+        assert (input.ndim == 2)
+        shape: Union[tuple[int, int], torch.Size] = input.shape
+        # For ROCm on MI300, the output fp8 dtype is torch.float_e3m3fnuz
```
Comment (jikunshang) [inline vllm/_ipex_ops.py]: can remove this?

**Review by jikunshang — APPROVED** [review]
Overall LGTM. thanks!

File: vllm/model_executor/layers/quantization/fp8.py [hunk 3]
```diff
@@ -203,7 +207,7 @@ def __init__(self, quant_config: Fp8Config):
         self.use_marlin = (not current_platform.has_device_capability(89)
                            or envs.VLLM_TEST_FORCE_FP8_MARLIN)
         # Disable marlin for rocm
-        if current_platform.is_rocm():
+        if current_platform.is_rocm() or current_platform.is_xpu():
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/fp8.py]: Should we say `not is_cuda` here?

```suggestion
        if not current_platform.is_cuda()
```

File: vllm/model_executor/layers/quantization/fp8.py [hunk 4]
```diff
@@ -776,6 +789,24 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
+                                  if self.block_quant else
+                                  layer.w13_weight_scale),
+                    w2_scale_inv=(layer.w2_weight_scale_inv if self.block_quant
+                                  else layer.w2_weight_scale),
+                    a1_scale_inv=layer.w13_input_scale,
+                    a2_scale_inv=layer.w2_input_scale,
+                    use_prepack=True,
+                )
+
+            return
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/fp8.py]: Is this a stray return that shouldn't be here?

File: vllm/model_executor/layers/quantization/fp8.py [hunk 5]
```diff
@@ -962,6 +993,25 @@ def apply(
         logical_to_physical_map: Optional[torch.Tensor] = None,
         logical_replica_count: Optional[torch.Tensor] = None,
     ) -> torch.Tensor:
+
+        if current_platform.is_xpu():
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/fp8.py]: cc @bnellnm @tlrmchlsmth @varun-sundar-rabindranath to review the Moe piece, this integration seems pretty intrusive

File: vllm/model_executor/layers/quantization/fp8.py [hunk 6]
```diff
@@ -440,6 +447,12 @@ def apply(self,
               layer: torch.nn.Module,
               x: torch.Tensor,
               bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        if current_platform.is_xpu():
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/fp8.py]: @LucasWilkinson another day without ScaledFP8Linear, another PR with more dispatching code...

File: vllm/model_executor/layers/quantization/fp8.py (hunk 5 above)
Comment (bnellnm) [reply vllm/model_executor/layers/quantization/fp8.py]: I think ideally there should be a XPU version of `Fp8MoEMethod` that is selected by `Fp8Config.get_quant_method` so that all the XPU specific bits could be isolated to that class.  But a similar kind of dispatching pattern is repeated in `modelopt.py` and `compressed_tensors_moe.py` so I'm not really opposed to this.

File: vllm/model_executor/layers/quantization/fp8.py [hunk 7]
```diff
@@ -137,10 +137,35 @@ def from_config(cls, config: dict[str, Any]) -> "Fp8Config":
...
+        elif isinstance(layer, Attention):
+            return Fp8KVCacheMethod(self)
+        return None
+
     def get_quant_method(self, layer: torch.nn.Module,
                          prefix: str) -> Optional["QuantizeMethodBase"]:
         from vllm.attention.layer import Attention  # Avoid circular import
 
+        if current_platform.is_xpu():
+            return self.get_xpu_quant_method(layer, prefix)
```
Comment (yma11) [inline vllm/model_executor/layers/quantization/fp8.py:168]: @ProExpertProg @bnellnm @LucasWilkinson Thanks for your review. I did a further refactor to avoid too many dispatches in current fp8 path and more future fp8 xpu impls. But since not like quantized models, there is no override mechanism for fp8 online quant so I still need such a router to xpu like here. Please take a review again.

**Comment by yma11** [comment]
@rogerxfeng8 @jikunshang FYI.

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 8]
```diff
@@ -251,3 +260,179 @@ def apply(self,
...
+
+class IPEXFp8LinearMethod(Fp8LinearMethod):
+
+    def __init__(self, quant_config: Fp8Config):
+        self.quant_config = quant_config
+        self.out_dtype = torch.get_default_dtype()
+        self.use_marlin = False
+        self.block_quant = self.quant_config.weight_block_size is not None
+
+    def create_weights(
```
Comment (jikunshang) [inline vllm/model_executor/layers/quantization/ipex_quant.py]: can remove this method?

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 9]
```diff
@@ -251,3 +260,179 @@ def apply(self,
         reshaped_x = x.reshape(-1, x.shape[-1])
         out = layer.ipex_qlinear(reshaped_x)
         return out.reshape(x.shape[:-1] + (layer.ipex_output_size, ))
+
+
+class IPEXFp8LinearMethod(Fp8LinearMethod):
```
Comment (jikunshang) [inline vllm/model_executor/layers/quantization/ipex_quant.py]: I prefer to use `XPUFp8LinearMethod` instead. same for moe method.

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 10]
```diff
@@ -251,3 +260,179 @@ def apply(self,
...
+            layer.weight = Parameter(qweight, requires_grad=False)
+            layer.weight_scale = Parameter(weight_scale, requires_grad=False)
+            layer.input_scale = None
+
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        weight = layer.weight.data
+        scale = layer.weight_scale.data
```
Comment (jikunshang) [inline vllm/model_executor/layers/quantization/ipex_quant.py]: weight_scale

File: vllm/platforms/xpu.py [hunk 11]
```diff
@@ -148,6 +148,10 @@ def get_current_memory_usage(cls,
         torch.xpu.reset_peak_memory_stats(device)
         return torch.xpu.max_memory_allocated(device)
 
+    @classmethod
+    def fp8_dtype(cls) -> torch.dtype:
```
Comment (jikunshang) [inline vllm/platforms/xpu.py:152]: add some comments why we support e5m2 not e4m3?

File: vllm/model_executor/layers/quantization/ipex_quant.py (hunk 8 above)
Comment (yma11) [reply vllm/model_executor/layers/quantization/ipex_quant.py]: yeah, I think so.

File: vllm/platforms/xpu.py (hunk 11 above)
Comment (yma11) [reply vllm/platforms/xpu.py:152]: will upstream that part in a following PR.

File: vllm/model_executor/layers/quantization/ipex_quant.py (hunk 9 above)
Comment (yma11) [reply vllm/model_executor/layers/quantization/ipex_quant.py]: updated.

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 12]
```diff
@@ -251,3 +260,165 @@ def apply(self,
...
+        self.topk_indices_dtype = None
+        self.layer = layer
+        self.quant_config = quant_config
+        self.block_quant = self.quant_config.weight_block_size is not None
+        self.fused_experts: Optional[
+            mk.FusedMoEModularKernel] = None  # type: ignore
+        self.use_marlin = False
+
+        # Check for DeepGemm support.
+        self.allow_deep_gemm = False
```
Comment (bnellnm) [inline vllm/model_executor/layers/quantization/ipex_quant.py]: Thanks for doing the refactoring!

Don't you need to call `super().__init__()` here?  (this will initialize `topk_indices_type`, `fused_experts` and `self.moe` There's also a bunch of member variables that don't seem necessary for this class, e.g. `use_marlin`, `allow_deep_gemm`.

Also, you might be able to inherit directly from `FusedMoEMethodBase` if you don't need any of the functionality from `Fp8MoEMethod`

Comment (yma11) [reply vllm/model_executor/layers/quantization/ipex_quant.py]: Yes, I can inherit directly from `FusedMoEMethodBase` and makes more sense by calling `super().__init__()`. Updated. Thanks.

**Review by bnellnm — APPROVED** [review]

File: vllm/_ipex_ops.py [hunk 13]
```diff
@@ -349,3 +350,55 @@ def copy_blocks(key_caches: list[torch.Tensor],
...
+        scale: Optional[torch.Tensor] = None,
+        num_token_padding: Optional[int] = None,
+        scale_ub: Optional[torch.Tensor] = None,
+        use_per_token_if_dynamic: bool = False,
+        output: Optional[torch.Tensor] = None,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """
+        Quantize input tensor to FP8 and return quantized tensor and scale.
+        
+        This function supports both static and dynamic quantization: If you
```
Comment (jikunshang) [inline vllm/_ipex_ops.py]: please update comments here. we should support dynamic quant currently.

Comment (yma11) [reply vllm/_ipex_ops.py]: updated.
